data-pipeline-flink/user-cache-updater/src/main/scala/org/sunbird/dp/usercache/functions/UserCacheUpdaterFunction.scala

```diff
@@ -0,0 +1,164 @@
+    }).getOrElse(metrics.incCounter(config.skipCount))
+  }
+
+
+  /**
+   * When Edata.State is Create/Created then insert the user data into redis in string formate
+   */
+  def createAction(userId: String, event: Event, metrics: Metrics): mutable.Map[String, AnyRef] = {
+    val userData: mutable.Map[String, AnyRef] = mutable.Map[String, AnyRef]()
+    Option(event.getContextDataId(cDataType = "SignupType")).map(signInType => {
+      if (config.userSelfSignedInTypeList.contains(signInType)) {
+        userData.put(config.userSignInTypeKey, config.userSelfSignedKey)
+      }
+      if (config.userValidatedTypeList.contains(signInType)) {
+        userData.put(config.userSignInTypeKey, config.userValidatedKey)
+      }
+    }).getOrElse(null)
+    userData
+  }
+
+  /**
+   * When edata.state is update/updated then update the user metadta
+   * information by reading from the cassandra
+   */
+
+  def updateAction(userId: String, event: Event, metrics: Metrics): mutable.Map[String, AnyRef] = {
+    val userCacheData: mutable.Map[String, AnyRef] = dataCache.getWithRetry(userId)
+
+    Option(event.getContextDataId(cDataType = "UserRole")).map(loginType => {
+      userCacheData.put(config.userLoginTypeKey, loginType)
+    })
+    val userMetaDataList = event.userMetaData()
+    if (!userMetaDataList.isEmpty() && userCacheData.contains(config.userSignInTypeKey) && ("Anonymous" != userCacheData.get(config.userSignInTypeKey))) {
+
+      // Get the user details from the cassandra table
+      val userDetails: mutable.Map[String, AnyRef] = userCacheData.++(extractUserMetaData(readFromCassandra(
+        keyspace = config.keySpace,
+        table = config.userTable,
+        QueryBuilder.eq("id", userId),
+        metrics)
+      ))
+
+      // Fetching the location details from the cassandra table
+      val updatedUserDetails: mutable.Map[String, AnyRef] = userDetails.++(extractLocationMetaData(readFromCassandra(
+        keyspace = config.keySpace,
+        table = config.locationTable,
+        clause = QueryBuilder.in("id", userDetails.get("locationids").getOrElse(new util.ArrayList()).asInstanceOf[util.ArrayList[String]]),
+        metrics)
+      ))
+      logger.info(s"User details ( $userId ) are fetched from the db's and updating the redis now.")
+      updatedUserDetails
+    } else {
+      logger.info(s"Skipping the event update from databases since event Does not have user properties or user sigin in type is Anonymous ")
+      userCacheData
+    }
+  }
+
+  def readFromCassandra(keyspace: String, table: String, clause: Clause, metrics: Metrics): util.List[Row] = {
+    var rowSet: util.List[Row] = null
+    val query = QueryBuilder.select.all.from(keyspace, table).where(clause).toString
+    try rowSet = cassandraConnect.find(query)
+    catch {
+      case ex: DriverException =>
+        cassandraConnect.reconnect()
+        rowSet = cassandraConnect.find(query)
+    }
+    metrics.incCounter(config.dbHitCount)
```

Comment:
What if the data is empty from cassandra? Is it still a dbhit? or should it be dbMiss?